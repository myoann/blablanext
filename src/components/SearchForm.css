.form {
  z-index: 3; /* Ensure the form is above the overlay */
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: 20px;
  width: 50%;
  padding: 10px 10px 10px 40px;
  background-color: rgba(255, 255, 255, 0.9);
  border-radius: 9999px;
}

.form label {
  width: 100%;
  font-weight: bold;
  color: #333;
}

.form > label > div:first-child {
  display: block;
  width: 90%;
  padding-top: 10px;
}

.form input {
  width: 100%;
  border: none;
  background-color: transparent;
}

.react-datepicker-popper {
  z-index: 999;
}

.form input:focus {
  outline: none;
}

.error {
  font-size: 10px;
  font-weight: 400;
  color: red;
}

.submitButton {
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 20px;
  background-color: #00aff5;
  color: white;
  border: none;
  border-radius: 50%;
  cursor: pointer;
  transition: background-color 0.5s;
}

.submitButton:disabled {
  background-color: #cccccc;
  cursor: not-allowed;
}

.submitButton:hover:not(:disabled) {
  background-color: #008ec2;
}

.searchIcon {
  filter: invert(100%);
}
