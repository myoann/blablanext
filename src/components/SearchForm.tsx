"use client";

import { useState } from "react";
import { createPortal } from "react-dom";
import Image from "next/image";
import DatePicker from "react-datepicker";

import { LatLng } from "@/types";

import PlacesAutocomplete from "./PlacesAutocomplete";

import "react-datepicker/dist/react-datepicker.css";
import "./SearchForm.css";

const todayDate = new Date().toISOString().substring(0, 10);

type FormErrors = {
  date?: string;
  fromCoordinate?: string;
  toCoordinate?: string;
};

type ValidateSearchForm = {
  fromCoordinate: LatLng | null;
  toCoordinate: LatLng | null;
  date: string | null;
};

/** Validate the form fields */
const validateSearchForm = ({
  fromCoordinate,
  toCoordinate,
  date,
}: ValidateSearchForm): FormErrors => {
  const errors: FormErrors = {};

  if (!fromCoordinate) {
    errors.fromCoordinate = "Enter a valid place";
  }

  if (!toCoordinate) {
    errors.toCoordinate = "Enter a valid place";
  }

  if (typeof date === "string") {
    const selectedDate = new Date(date).toISOString().substring(0, 10);

    if (selectedDate < todayDate) {
      errors.date = "Select a date in the future";
    } else {
      const oneYearFromNow = new Date();
      oneYearFromNow.setFullYear(oneYearFromNow.getFullYear() + 1);
      const oneYearFromNowDate = oneYearFromNow.toISOString().substring(0, 10);

      if (selectedDate > oneYearFromNowDate) {
        errors.date = "Select a date within the next year";
      }
    }
  }

  return errors;
};

const SearchForm = () => {
  const [fromCoordinate, setFromCoordinate] = useState<LatLng | null>(null);
  const [toCoordinate, setToCoordinate] = useState<LatLng | null>(null);
  const [startDate, setStartDate] = useState<Date | null>(new Date());

  const [isPending, setIsPending] = useState(false);
  const [errors, setErrors] = useState<FormErrors>({});

  const handleFromCoordinates = (coords: LatLng) => {
    setFromCoordinate(coords);
  };

  const handleToCoordinates = (coords: LatLng) => {
    setToCoordinate(coords);
  };

  const handleFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    setErrors({});
    setIsPending(true);

    const date = startDate?.toISOString().substring(0, 10) || null;

    const validationErrors = validateSearchForm({
      fromCoordinate,
      toCoordinate,
      date,
    });

    setIsPending(false);

    if (Object.keys(validationErrors).length > 0) {
      setErrors(validationErrors);
      return;
    }

    window.location.href = `/search?fc=${fromCoordinate?.lat},${fromCoordinate?.lng}&tc=${toCoordinate?.lat},${toCoordinate?.lng}&db=${date}`;
  };

  return (
    <form
      onSubmit={handleFormSubmit}
      aria-label="Search for a carpool"
      className="form"
    >
      <label>
        From coordinate
        <PlacesAutocomplete
          onCoordinates={handleFromCoordinates}
          placeholder="Enter a departure place"
        />
        {errors.fromCoordinate && (
          <span className="error">{errors.fromCoordinate}</span>
        )}
      </label>

      <label>
        To coordinate
        <PlacesAutocomplete
          onCoordinates={handleToCoordinates}
          placeholder="Enter a destination place"
        />
        {errors.toCoordinate && (
          <span className="error">{errors.toCoordinate}</span>
        )}
      </label>

      <label>
        Date
        <DatePicker
          selected={startDate}
          onChange={(date) => setStartDate(date)}
          popperContainer={({ children }) => {
            if (typeof document !== "undefined") {
              return createPortal(children, document.body);
            }
            return null;
          }}
          required
        />
        {errors.date && <span className="error">{errors.date}</span>}
      </label>

      <button type="submit" disabled={isPending} className="submitButton">
        <Image
          src="/search.svg"
          alt="Search"
          width={24}
          height={24}
          className="searchIcon"
        />
      </button>
    </form>
  );
};

export default SearchForm;
