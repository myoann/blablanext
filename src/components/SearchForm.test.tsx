import { render } from "@testing-library/react";

import SearchForm from "./SearchForm";

describe("SearchForm", () => {
  it("matches the snapshot", () => {
    const { container } = render(<SearchForm />);

    expect(container).toMatchSnapshot();
  });

  it("renders a form", () => {
    const { container } = render(<SearchForm />);

    expect(container.querySelector("form")).toBeInTheDocument();
  });

  it("renders an input for the from coordinate", () => {
    const { container } = render(<SearchForm />);

    expect(
      container.querySelector('input[name="fromCoordinate"]'),
    ).toBeInTheDocument();
  });

  it("renders an input for the to coordinate", () => {
    const { container } = render(<SearchForm />);

    expect(
      container.querySelector('input[name="toCoordinate"]'),
    ).toBeInTheDocument();
  });

  it("renders an input for the date", () => {
    const { container } = render(<SearchForm />);

    expect(container.querySelector('input[type="date"]')).toBeInTheDocument();
  });

  it("renders a button", () => {
    const { container } = render(<SearchForm />);

    expect(container.querySelector("button")).toBeInTheDocument();
  });

  it("renders an image inside the button", () => {
    const { container } = render(<SearchForm />);

    expect(container.querySelector("button img")).toBeInTheDocument();
  });
});
