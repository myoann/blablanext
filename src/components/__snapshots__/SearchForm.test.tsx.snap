// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm matches the snapshot 1`] = `
<div>
  <form
    aria-label="Search for a carpool"
    class="searchForm"
  >
    <label>
      From city
      <div
        class="placesAutocomplete"
      >
        <input
          placeholder="Enter a departure place"
          value=""
        />
        <ul />
      </div>
    </label>
    <label>
      To city
      <div
        class="placesAutocomplete"
      >
        <input
          placeholder="Enter a destination place"
          value=""
        />
        <ul />
      </div>
    </label>
    <label>
      Date
      <input
        type="date"
      />
    </label>
    <button
      class="submitButton"
      type="submit"
    >
      <img
        alt="Search"
        class="searchIcon"
        data-nimg="1"
        decoding="async"
        height="24"
        loading="lazy"
        src="/search.svg"
        style="color: transparent;"
        width="24"
      />
    </button>
  </form>
</div>
`;
